### ISSUE TITLE
@abstractmethod should only be allowed on classes derived from ABCMeta

### ISSUE THREAD

**Issue by rchiodo** [microsoft/pyright#3952]
**Is your feature request related to a problem? Please describe.**
A clear and concise description of the problem.

This code here is not flagged in any way by pyright:

```python
class Z():
    @abstractmethod
    def foo(self):
        print('z foo')

class M(Z):
    pass


x = M()
x.foo()
```

**Describe the solution you'd like**
The reason this code isn't flagged is because the `Z` class doesn't derive from ABCMeta, but it's using @abstractmethod. 

It would be nice if the definition of the class `Z` flagged a warning or error that it should be derived from ABCMeta, as described in the python docs (which say it's necessary in order to use @abstractmethod)
https://docs.python.org/3/library/abc.html#abc.abstractmethod

**Additional context**

This code is flagged correctly:

```python
class Z(ABC):
    @abstractmethod
    def foo(self):
        print('z foo')

class M(Z):
    pass


x = M()
x.foo()
```

Because pyright can now know that Z is actually an abstract class.

**Comment by erictraut** [comment]
I'm reluctant to make this an error in pyright. It doesn't cause a runtime exception, and it's not a type violation. Arguably, it's a linter-style error, something more appropriate for pylint to detect.

**Comment by rchiodo** [comment]
It could cause unexpected behavior though? Suppose Z was like this because it was supposed to be overridden:

```
class Z():
    @abstractmethod
    def foo(self):
        raise Except('Method needs to be overridden')
```

Pyright not detecting the `@abstractmethod` could cause unexpected runtime failures.

**Comment by erictraut** [comment]
Also, I think mypy treats a class as "effectively abstract" if one or more methods are decorated with `@abstractmethod`. There is no runtime enforcement without the `ABCMeta` metaclass, but mypy applies its abstract class checks in this case. Pyright doesn't make the same assumption because this approach imposes a high performance cost for all classes (most of which are not abstract and don't use `@abstractmethod`) for a lazy evaluator like pyright. If someone uses both mypy and pyright (which is not uncommon), we would likely get negative feedback if pyright were to flag the use of `@abstractmethod` as an error in a non-`ABC` class.

At runtime, `@abstractmethod` is a no-op if the class doesn't derive from `ABC`, and that's how pyright treats it. So I don't see any potential for an unexpected runtime failure.

**Comment by rchiodo** [comment]
The unexpected runtime failure stems from the user forgetting to subclass from ABC but they didn't. 

I'm not suggesting pyright check all uses of classes for potentially being abstract, but rather check marked functions in class as abstract as actually deriving from ABCmeta. Which I think can be done once when binding the class? Maybe that's expensive too.

**Comment by rchiodo** [comment]
I think you're saying the 'forgetting' to subclass ABC might be on purpose, so why flag it?
